### PR TITLE
Allow a do_last test target

### LINKED ISSUE #15230: Add a prereq option "do_last" to run_tests
## Reason
I currently have a lot of tests in a spec file with a RunCommand for a postprocessing script. The list for the RunCommand `prereq =` is getting quite long.

## Design
If there was a `do_last` or similar option it would save a bunch of copy/paste. @milljm suggested `--pedantic` but I would also like to still be able to submit jobs in parallel.

## Impact
Save some time and possibly mistakes in building more complicated spec files

### PR DESCRIPTION
**Pull request by milljm**
For test spec files with many tests, and those wishing to have a final 'do this last' test,
users can provide the 'prereq = ALL' param in their test spec files:

Closes #15230

### CODE REVIEW

**Comment by permcody** [comment]
@milljm - This PR has conflicts but I want to make sure that this feature is really necessary. I suspect that you don't really need it, but it's a convenience feature? You could use the `prereq` flag to depend on all of the other tests right? Since you are going to have to redo things, maybe you should consider using `prereq = ALL` or something. I think you can already use "ALL" for other caveats like compilers and platforms so it would be more similar to what we already have, also since this parameter by it's very definition "over prescribes" conditions, there are many weird combinations that you might have to think about. What happens when a user puts in prereq and do_last, or any combination of those throughout the file. If you condensed this capability down to a single parameter there would be no issues with combinations of parameters. 

For a final example, think about Andrew's block and boundary restrictable classes in MOOSE. You can say an object applies to "ALL" boundaries, or blocks. It can be very handy.

**Comment by loganharbour** [comment]
...how's this lookin @milljm?

**Comment by cticenhour** [comment]
@permcody - @milljm and I were talking about this PR late last week, and were trying to go through a few permutations of what users might have in their various test configurations (duplicates of `do_last` and `prereq`, `do_last` in the super block, but conflicting `prereq` in the test sub-block, etc.). The DAG configuration that's being used here is robust against adding duplicate "edges" (or, the same relationship previously established by another parameter setting) without error, thereby lowering the risk of weird combinations that the user might find themselves in. I was going to add more tests to this today. 

I recognize that there may yet be some corner cases we didn't think of, but we were also concerned that something like simply `prereq = ALL` might run us into a naming conflict with a test. Maybe making it more specific and painfully obvious, like `ALL_OTHER_TESTS`?

**Comment by permcody** [comment]
@cticenhour - I still like ALL because it's consistent with MOOSE. The more we have similar options throughout the better. If somebody screws up, no big deal! We could easily add a warning for when somebody names a test "ALL".

**Comment by cticenhour** [comment]
That's a fair point - do we have a `mooseWarning` equivalent for our Python scripts?

**Comment by cticenhour** [comment]
@permcody @milljm - I plan to squash this back down into one commit, but I'm keeping it separate for now so that you can comment specifically about what I've added after the original commit. `prereq = ALL` now performs the same function as `do_last = true`, and an extra test has been added to check the functionality of the case where any other test name in the spec file is "ALL" when `prereq = ALL`, which throws a warning (a newly created Status) when this occurs. If desired, I can rollback to 9407732, where this was treated as a full error. 

A few notes:

- `mooseWarning` didn't associate itself with the test printout when it was thrown, so a new status seemed like a better option.
- Similar to other warnings in MOOSE, a warning status doesn't fail the test (exit code 0), but here, the test itself does not complete and a notice of this is displayed to the user.
- I did experiment with making this a non-zero exit code, so we could use it in CIVET. But, I wasn't sure if that behavior was desired, considering other warnings in MOOSE don't cause an immediate failure in the TestHarness. I'd like some feedback on this point. 

Here's the current behavior on my machine with `MOOSE_TERM_FORMAT=tpsnc`
```
 WARNING TEST DID NOT RUN (Test named ALL when "prereq = ALL" elsewhere in test spec file) tests/test_harness.ALL
      OK tests/test_harness.do_last
--------------------------------------------------------------------------------------------------------------
Ran 1 tests in 0.4 seconds.
1 passed, 1 warnings, 0 skipped, 0 pending, 0 failed
```

**Review by permcody — DISMISSED** [review]

**Comment by permcody** [comment]
This looks much better!

**Comment by cticenhour** [comment]
Awesome - do you have any comments about the new warning status, or how we should handle it as far as exit code?

**Comment by permcody** [comment]
I think I'd prefer an error since that's what we do with bad prereqs now. The problem with warnings is that they get ignored. This is especially true if your exit code is zero and the targets are green in CIVET.

File: python/TestHarness/testers/Tester.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def validParams():
 
         params.addParam('heavy',    False, "Set to True if this test should only be run when the '--heavy' option is used.")
         params.addParam('group',       [], "A list of groups for which this test belongs.")
-        params.addParam('prereq',      [], "A list of prereq tests that need to run successfully before launching this test.")
+        params.addParam('prereq',      [], "A list of prereq tests that need to run successfully before launching this test. When 'prereq = ALL', TestHarness will run this test last. Multiple 'prereq = ALL' tests, or tests that depend on a 'prereq = ALL' test will result in cyclic errors. Naming a test 'ALL' when using 'prereq = ALL' will also result in an error.")
```
Comment (cticenhour) [inline python/TestHarness/testers/Tester.py:35]: Found a typo - 
```suggestion
        params.addParam('prereq',      [], "A list of prereq tests that need to run successfully before launching this test. When 'prereq = ALL', TestHarness will run this test last. Multiple 'prereq = ALL' tests, or tests that depend on a 'prereq = ALL' test will result in cyclic errors. Naming a test 'ALL' when using 'prereq = ALL' will also result in a warning.")
```

**Comment by cticenhour** [comment]
Sounds good - I'll roll this back to an error then. Thanks!

**Comment by cticenhour** [comment]
Alright @permcody. The current behavior is: 

```
    SKIP tests/test_harness.do_last [skipped dependency]
tests/test_harness.ALL: Working Directory: /Users/icenct/projects/moose/test/tests/test_harness
tests/test_harness.ALL: Running command: /Users/icenct/projects/moose/test/moose_test-devel -i good.i --error --error-unused --error-override --no-gdb-backtrace
tests/test_harness.ALL:
   ERROR FAILED (Test named ALL when "prereq = ALL" elsewhere in test spec file!) tests/test_harness.ALL


Final Test Results:
--------------------------------------------------------------------------------------------------------------
    SKIP tests/test_harness.do_last [skipped dependency]
   ERROR FAILED (Test named ALL when "prereq = ALL" elsewhere in test spec file!) tests/test_harness.ALL
--------------------------------------------------------------------------------------------------------------
Ran 1 tests in 0.2 seconds.
0 passed, 1 skipped, 0 pending, 1 FAILED
```

**Comment by cticenhour** [comment]
@gardnerru Does this get you what you need?

**Comment by milljm** [comment]
@cticenhour thank you for working on this!

**Comment by cticenhour** [comment]
@milljm Learning how the Job DAGs worked was kinda fun! 😄

**Comment by gardnerru** [comment]
@cticenhour Glad you had fun. Ill send all my ridiculous requests to you now ;)

Looks like a winner to me.

**Comment by cticenhour** [comment]
@permcody I think this is ready for a final approval now that I've cleaned up the commit history.

**Review by permcody — APPROVED** [review]
Looks good - nice tests!

**Pull request merged**